docs/src/tutorial-nondeterministic-variables.md

```diff
@@ -1,95 +1,103 @@
- 2. Specify invariants for types that you define, enabling the creation of safe nondeterministic variables for those types.
+In other words, `kani::any()` should not produce values that are invalid for the type.
 
 ## Safe nondeterministic variables
 
-Let's say you're developing an inventory management tool, and you would like to verify that your API to manage items is correct.
-Here is a simple implementation of this API:
+Let's say you're developing an inventory management tool, and you would like to start verifying properties about your API.
+Here is a simple example (available [here](https://github.com/model-checking/kani/blob/main/docs/src/tutorial/arbitrary-variables/src/inventory.rs)):
 
 ```rust
 {{#include tutorial/arbitrary-variables/src/inventory.rs:inventory_lib}}
 ```
 
-Now we would like to verify that, no matter which combination of `id` and `quantity`, a call to `Inventory::update()` followed by a call to `Inventory::get()` using the same `id` returns some value that's equal to the one we inserted:
+Let's write a fairly simple proof harness, one that just ensures we successfully `get` the value we inserted with `update`:
 
 ```rust
 {{#include tutorial/arbitrary-variables/src/inventory.rs:safe_update}}
 ```
 
-In this harness, we use `kani::any()` to generate the new `id` and `quantity`.
-`kani::any()` is a **safe** API function, and it represents only valid values.
+We use `kani::any()` twice here:
 
-If we run this example, Kani verification will succeed, including the assertion that shows that the underlying `u32` variable  used to represent `NonZeroU32` cannot be zero, per its type invariant:
+1. `id` has type `ProductId` which was actually just a `u32`, and so any value is fine.
+2. `quantity`, however, has type `NonZeroU32`.
+In Rust, it would be undefined behavior to have a value of `0` for this type.
 
-You can try it out by running the example under
-[`arbitrary-variables`](https://github.com/model-checking/kani/tree/main/docs/src/tutorial/arbitrary-variables/):
+We included an extra assertion that the value returned by `kani::any()` here was actually non-zero.
+If we run this, you'll notice that verification succeeds.
 
-```
+```bash
 cargo kani --harness safe_update
 ```
 
-## Unsafe nondeterministic variables
+`kani::any()` is safe Rust, and so Kani only implements it for types where type invariants are enforced.
+For `NonZeroU32`, this means we never return a `0` value.
+The assertion we wrote in this harness was just an extra check we added to demonstrate this fact.
 
-Kani also includes an **unsafe** method to generate unconstrained nondeterministic variables which do not take type invariants into consideration.
-As with any unsafe method in Rust, users have to make sure the right guardrails are put in place to avoid undesirable behavior.
+## Other kinds of nondeterministic variables
```

Comment:
What are you trying to say in this section? I'm tempted to just remove this for now.

Reply:
I think this structure is good bones, but I see what you're saying about it seeming pointless right now.

As soon as we decide things like `any_vec` are stable and we commit to their API, this is where I'd put it. So this section was kinda meant to be about that until I went and audited the docs and felt that maybe we wanted to revise that API first...

Reply:
I agree this section is a bit confusing. Not sure what it's getting at.